drivers/network/tcpip/tcpip/icmp.c

```diff
@@ -0,0 +1,389 @@
+
+        return Status;
+    }
+
+    KeQueryPerformanceCounter(&TimerResolutionSec);
+
+    AddrFile = (PADDRESS_FILE)SendContext->TdiRequest.Handle.AddressHandle;
+
+    // setting up the context
+
+    SendContext->Irp = Irp;
+    SendContext->TimerResolution = TimerResolutionSec.QuadPart / 1000; // now in milliseconds
+    SendContext->CurrentReply = Irp->AssociatedIrp.SystemBuffer;
+    SendContext->RemainingSize = ReplySize;
+    SendContext->nReplies = 0;
+    SendContext->ErrorStatus = STATUS_SUCCESS;
+    KeInitializeSpinLock(&SendContext->ProcessingLock);
+    KeInitializeEvent(&SendContext->InitializationFinishedEvent, SynchronizationEvent, FALSE);
+
+    KeInitializeDpc(&TimeoutDpc, &EndRequestHandler, SendContext);
+
+    LockObject(AddrFile, &OldIrql);
+
+    AddrFile->TTL = Request->Ttl;
+
+    AddrFile->ReceiveDatagramHandlerContext = SendContext;
+    AddrFile->ReceiveDatagramHandler = ReceiveDatagram;
+    AddrFile->RegisteredReceiveDatagramHandler = TRUE;
+
+    ConnectionInfo.RemoteAddress = &RemoteAddressTa;
+    ConnectionInfo.RemoteAddressLength = sizeof(RemoteAddressTa);
+
+    UnlockObject(AddrFile, OldIrql);
+
+    RequestSize = sizeof(ICMP_HEADER) + Request->DataSize;
+
+    // making up the request packet
+    Buffer = ExAllocatePoolWithTag(NonPagedPool, RequestSize, OUT_DATA_TAG);
+
+    if (!Buffer)
+    {
+        NTSTATUS _Status = FileCloseAddress(&SendContext->TdiRequest);
+        ASSERT(NT_SUCCESS(_Status));
+
+        ExFreePoolWithTag(SendContext, OUT_DATA_TAG);
+
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    ((PICMP_HEADER)Buffer)->Type = ICMP_TYPE_ECHO_REQUEST;
+    ((PICMP_HEADER)Buffer)->Code = ICMP_TYPE_ECHO_REPLY;
+    ((PICMP_HEADER)Buffer)->Checksum = 0;
+    ((PICMP_HEADER)Buffer)->Identifier = (UINT_PTR)PsGetCurrentProcessId() & UINT16_MAX;
+    ((PICMP_HEADER)Buffer)->Seq = InterlockedIncrement16(&IcmpSequence);
+    memcpy(Buffer + sizeof(ICMP_HEADER), (PUCHAR)Request + Request->DataOffset, Request->DataSize);
+    ((PICMP_HEADER)Buffer)->Checksum = IPv4Checksum(Buffer, RequestSize, 0);
+
+    Status = AddrFile->Send(
+        AddrFile,
+        &ConnectionInfo,
+        (PCHAR)Buffer,
+        RequestSize,
+        &DataUsed);
+
+    // From this point we may receive a reply packet.
+    // But we are not ready for it thus InitializationFinishedEvent is needed (see below)
+
+    SendContext->StartTicks = KeQueryPerformanceCounter(NULL).QuadPart;
+
+    ExFreePoolWithTag(Buffer, OUT_DATA_TAG);
+
+    if (!NT_SUCCESS(Status))
+    {
+        NTSTATUS _Status = FileCloseAddress(&SendContext->TdiRequest);
+        ASSERT(NT_SUCCESS(_Status));
+
+        ExFreePoolWithTag(SendContext, OUT_DATA_TAG);
+
+        TI_DbgPrint(DEBUG_ICMP, ("Failed to send a datagram: 0x%x\n", Status));        
+        return Status;
+    }
+
+    // reusing the LARGE_INTEGER variable
+    TimerResolutionSec.QuadPart = (-1LL) * 10 * 1000 * Request->Timeout;
```

Comment:
Since `Request` points to the IRP's `SystemBuffer`, `Request->Timeout` may have already been overwritten with output data at this point. So you need to save the value before calling `Send()`.